File: sdk-py/noether_sdk/transport.py
"""Shared HTTP transport for every sub-client."""

from __future__ import annotations
import time
from dataclasses import dataclass
from typing import Any

import httpx

from .errors import NetworkError, classify_error


@dataclass
class Credentials:
    key_id: str
    secret: str


class Transport:
    """Async HTTP client that handles Bearer auth + error classification."""

    def __init__(
        self,
        base_url: str,
        *,
        client: httpx.AsyncClient | None = None,
        timeout: float = 30.0,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self._owns_client = client is None
        self._client = client or httpx.AsyncClient(timeout=timeout)

    async def aclose(self) -> None:
        if self._owns_client:
            await self._client.aclose()

    async def request(
        self,
        method: str,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        json: Any | None = None,
        credentials: Credentials | None = None,
        with_timestamp: bool = True,
    ) -> Any:
        url = f"{self.base_url}{path}" if path.startswith("/") else f"{self.base_url}/{path}"
        headers: dict[str, str] = {"accept": "application/json"}
        if json is not None:
            headers["content-type"] = "application/json"
        if credentials is not None:
            headers["authorization"] = f"Bearer {credentials.key_id}:{credentials.secret}"
            if with_timestamp:
                headers["x-timestamp"] = str(int(time.time()))

        clean_params = (
            {k: v for k, v in params.items() if v is not None} if params else None
        )

        try:
            response = await self._client.request(
                method,
                url,
                params=clean_params,
                json=json,
                headers=headers,
            )
        except httpx.RequestError as exc:
            raise NetworkError(f"{method} {url} failed: {exc}", cause=exc) from exc

        body: dict[str, Any] | None = None
        try:
            body = response.json() if response.text else None
        except ValueError:
            body = None

        if not response.is_success:
            retry_after_raw = response.headers.get("retry-after")
            retry_after = int(retry_after_raw) if retry_after_raw and retry_after_raw.isdigit() else None
            raise classify_error(response.status_code, body, url, retry_after)
        return body

File: sdk-ts/src/transport.ts
import { ApiErrorBody, NetworkError, classifyError } from './errors.js';

export interface Credentials {
  keyId: string;
  secret: string;
}

export interface TransportOptions {
  baseUrl: string;
  fetch?: typeof fetch;
  defaultHeaders?: Record<string, string>;
}

export interface RequestOptions {
  method?: 'GET' | 'POST' | 'DELETE';
  path: string;
  query?: Record<string, string | number | boolean | undefined>;
  body?: unknown;
  credentials?: Credentials | null;
  /** Send X-Timestamp header for replay protection (default true when authed). */
  withTimestamp?: boolean;
  signal?: AbortSignal;
}

const RETRY_AFTER_HEADER = 'retry-after';

export class Transport {
  private readonly baseUrl: string;
  private readonly fetchImpl: typeof fetch;
  private readonly defaultHeaders: Record<string, string>;

  constructor(opts: TransportOptions) {
    this.baseUrl = opts.baseUrl.replace(/\/$/, '');
    this.fetchImpl = opts.fetch ?? globalThis.fetch;
    if (!this.fetchImpl) {
      throw new Error('global fetch is unavailable; pass `fetch` in NoetherClient options (Node 18+ has it native).');
    }
    this.defaultHeaders = opts.defaultHeaders ?? {};
  }

  async request<T = unknown>(opts: RequestOptions): Promise<T> {
    const url = this.buildUrl(opts.path, opts.query);
    const headers: Record<string, string> = {
      accept: 'application/json',
      ...this.defaultHeaders,
    };

    if (opts.body !== undefined) {
      headers['content-type'] = 'application/json';
    }

    if (opts.credentials) {
      headers.authorization = `Bearer ${opts.credentials.keyId}:${opts.credentials.secret}`;
      if (opts.withTimestamp !== false) {
        headers['x-timestamp'] = String(Math.floor(Date.now() / 1000));
      }
    }

    let response: Response;
    try {
      response = await this.fetchImpl(url, {
        method: opts.method ?? 'GET',
        headers,
        body: opts.body !== undefined ? JSON.stringify(opts.body) : undefined,
        signal: opts.signal,
      });
    } catch (err) {
      throw new NetworkError(`fetch ${opts.method ?? 'GET'} ${url} failed`, err);
    }

    const text = await response.text();
    let parsed: unknown = null;
    if (text.length > 0) {
      try {
        parsed = JSON.parse(text);
      } catch {
        // non-JSON; surface raw text below
      }
    }

    if (!response.ok) {
      const body = (typeof parsed === 'object' && parsed !== null ? (parsed as ApiErrorBody) : { error: text || `HTTP ${response.status}` });
      const retryHeader = response.headers.get(RETRY_AFTER_HEADER);
      const retryAfter = retryHeader ? Number(retryHeader) : null;
      throw classifyError(response.status, body, url, Number.isFinite(retryAfter) ? retryAfter : null);
    }

    return parsed as T;
  }

  private buildUrl(path: string, query?: RequestOptions['query']): string {
    const url = new URL(path.startsWith('/') ? `${this.baseUrl}${path}` : `${this.baseUrl}/${path}`);
    if (query) {
      for (const [k, v] of Object.entries(query)) {
        if (v === undefined) continue;
        url.searchParams.set(k, String(v));
      }
    }
    return url.toString();
  }
}
